### PR TITLE
refactor(protocol): model ResponseTool as tagged enum to match Responses spec and tighten MCP validation

### PR DESCRIPTION
**Pull request by zhaowenzi**
### Problem

ResponseTool was previously modeled as a flattened struct with a type field plus many optional fields shared across tool kinds. This made it easy for invalid combinations (e.g. MCP-only fields on non-MCP tools) to silently pass parsing/validation, and it doesn’t scale well as OpenAI’s Responses API has (we are going to support) more tool types and tool-specific fields. Extending support would require more and more ad-hoc, per-field validation rules to prevent “field bleed”.

### Solution

Refactor ResponseTool to match the OpenAI Responses API tool schema: an internally tagged enum (type-discriminated) where each tool type owns only its valid fields. This shifts correctness from “runtime validation of a permissive struct” to “type-level correctness”, reducing validation complexity and making it straightforward to add additional OpenAI tool types in the future without proliferating conditional checks.

## Changes

- Refactored ResponseTool from a flat struct to #[serde(tag = "type")] enum with tool-specific structs (FunctionTool, McpTool, WebSearchPreviewTool, CodeInterpreterTool).
- Updated tool extraction and Harmony/OpenAI routers to work with the new representation; MCP tool-call classification uses session.has_exposed_tool() (MCP tools are exposed as function tools to the model).
- Updated tests to construct tools using the new enum variants; added a deserialization test to ensure unknown fields on function tools are rejected.

## Test Plan

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated

</details>

## Release Notes

* **Refactor**
  * Restructured internal tool representation system for improved consistency and maintainability across tool types (function, MCP, web search, and code interpreter).
  * Enhanced validation for MCP tools, now enforcing required server label configuration and stricter format constraints.
  * Simplified tool selection logic to rely on session-based exposure checks rather than precomputed tool sets.

### CODE REVIEW

File: model_gateway/src/routers/mcp_utils.rs [hunk 1]
```diff
@@ -254,16 +253,15 @@ pub async fn ensure_request_mcp_client(
...
-            headers: tool.headers.clone().unwrap_or_default(),
-            allowed_tools: tool.allowed_tools.clone(),
+        .filter_map(|tool| match tool {
+            ResponseTool::Mcp(mcp) => Some(McpServerInput {
+                label: mcp.server_label.clone(),
+                url: mcp.server_url.clone(),
+                authorization: mcp.authorization.clone(),
+                headers: mcp.headers.clone().unwrap_or_default(),
+                allowed_tools: mcp.allowed_tools.clone(),
+            }),
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply model_gateway/src/routers/mcp_utils.rs:263]: This SSRF risk already existed before and was not introduced by this PR. Keep it for now

File: protocols/src/responses.rs [hunk 2]
```diff
@@ -1162,53 +1176,42 @@ fn validate_response_tools(tools: &[ResponseTool]) -> Result<(), ValidationError
...
+                    format!("Duplicate MCP server_label '{raw_label}' found in 'tools' parameter.")
                         .into(),
-                    );
-                    return Err(e);
-                }
+                );
+                return Err(e);
             }
-            _ => {}
         }
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply protocols/src/responses.rs:1215]: This is a pre-existing issue and not introduced by this PR. Keep it for now

File: model_gateway/src/routers/grpc/harmony/responses/non_streaming.rs [hunk 3]
```diff
@@ -168,12 +167,12 @@ async fn execute_with_mcp_loop(
...
-                // Separate MCP and function tool calls based on tool type
-                let request_tools = current_request.tools.as_deref().unwrap_or(&[]);
-                let mcp_tool_names = build_mcp_tool_names_set(request_tools);
+                // Separate MCP and function tool calls based on session exposure.
+                // MCP tools are exposed to the model as function tools, so the only reliable
+                // discriminator is whether the name belongs to the MCP session.
                 let (mcp_tool_calls, function_tool_calls): (Vec<_>, Vec<_>) = tool_calls
                     .into_iter()
-                    .partition(|tc| mcp_tool_names.contains(tc.function.name.as_str()));
+                    .partition(|tc| session.has_exposed_tool(&tc.function.name));
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply model_gateway/src/routers/grpc/harmony/responses/non_streaming.rs:176]: This naming conflict isn’t new. The right fix is to handle it when building exposed tool names (reserve user function tool names too), not only during MCP partitioning; we’ll address it in a follow-up.

File: model_gateway/src/routers/openai/responses/utils.rs [hunk 4]
```diff
@@ -208,28 +208,28 @@ pub(super) fn insert_optional_value<T: Serialize>(
...
                     Value::Array(allowed.iter().map(|s| json!(s)).collect()),
                 );
             }
             Some(Value::Object(m))
         }
-        ResponseToolType::WebSearchPreview => Some(json!({"type": "web_search_preview"})),
-        ResponseToolType::CodeInterpreter => Some(json!({"type": "code_interpreter"})),
+        ResponseTool::WebSearchPreview(_) => serde_json::to_value(tool).ok(),
+        ResponseTool::CodeInterpreter(_) => serde_json::to_value(tool).ok(),
         _ => None,
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply model_gateway/src/routers/openai/responses/utils.rs]: Done

File: mcp/src/responses_bridge.rs [hunk 5]
```diff
@@ -88,21 +88,15 @@ pub fn build_response_tools_with_names(
...
-            allowed_tools: None,
+        .map(|entry| {
+            ResponseTool::Function(FunctionTool {
+                function: Function {
+                    name: resolved_name_for_entry(entry, exposed_names).to_string(),
+                    description: entry.tool.description.as_ref().map(|d| d.to_string()),
+                    parameters: Value::Object((*entry.tool.input_schema).clone()),
+                    strict: None,
+                },
+            })
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply mcp/src/responses_bridge.rs:99]: Done

File: model_gateway/src/routers/openai/responses/utils.rs [hunk 6]
```diff
@@ -208,28 +208,28 @@ pub(super) fn insert_optional_value<T: Serialize>(
...
                     "allowed_tools".to_string(),
                     Value::Array(allowed.iter().map(|s| json!(s)).collect()),
                 );
             }
             Some(Value::Object(m))
         }
-        ResponseToolType::WebSearchPreview => Some(json!({"type": "web_search_preview"})),
-        ResponseToolType::CodeInterpreter => Some(json!({"type": "code_interpreter"})),
+        ResponseTool::WebSearchPreview(_) => serde_json::to_value(tool).ok(),
+        ResponseTool::CodeInterpreter(_) => serde_json::to_value(tool).ok(),
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply model_gateway/src/routers/openai/responses/utils.rs:232]: N/A

File: model_gateway/src/routers/grpc/harmony/streaming.rs [hunk 7]
```diff
@@ -743,20 +731,14 @@ impl HarmonyStreamingProcessor {
...
+                                    if s.has_exposed_tool(tool_name) {
+                                        Some(s.tool_response_format(tool_name))
                                     } else {
-                                        None // Function tool
+                                        None
                                     }
-                                } else {
-                                    None // No MCP context, treat as function tool
-                                };
+                                });
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply model_gateway/src/routers/grpc/harmony/streaming.rs:741]: n/a

File: protocols/src/responses.rs [hunk 8]
```diff
@@ -20,41 +20,64 @@ use crate::{builders::ResponsesResponseBuilder, validated::Normalizable};
...
-    pub r#type: ResponseToolType,
-    // Function tool fields (used when type == "function")
-    // In Responses API, function fields are flattened at the top level
+#[serde(deny_unknown_fields)]
+pub struct FunctionTool {
+    /// Flatten to match Responses API tool JSON shape.
     #[serde(flatten)]
-    pub function: Option<Function>,
-    // MCP-specific fields (used when type == "mcp")
+    pub function: Function,
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply protocols/src/responses.rs:50]: We keep flatten to match the Responses API JSON shape (also reuse the function from common) and keep deny_unknown_fields for strictness. To avoid relying on assumptions about serde edge cases, we added a deserialization test that confirms unknown fields on type=function tools are rejected.
